etcd-backup/main.go

```diff
@@ -0,0 +1,116 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/spf13/cobra"
+	"go.etcd.io/etcd/client/pkg/v3/transport"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/client/v3/snapshot"
+
+	"go.uber.org/zap"
+)
+
+const (
+	DefaultDialTimeout       = 15 * time.Second
+	DefaultEtcdClientTimeout = 5 * time.Minute
+)
+
+type options struct {
+	clusterID      string
+	etcdEndpoint   string
+	s3BucketName   string
+	s3BucketRegion string
+}
+
+func main() {
+	var opts options
+
+	cmd := &cobra.Command{
+		Use:          "etcd-backup",
+		SilenceUsage: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return run(opts)
+		},
+	}
+
+	cmd.Flags().StringVar(&opts.clusterID, "cluster-id", "", "ID of the hosted cluster, used as a key for the snapshot file in S3")
+	cmd.Flags().StringVar(&opts.etcdEndpoint, "etcd-endpoint", "", "endpoint of the etcd cluster to backup.")
+	cmd.Flags().StringVar(&opts.s3BucketName, "s3-bucket-name", "", "name of the S3 bucket to store etcd backups.")
+	cmd.Flags().StringVar(&opts.s3BucketRegion, "s3-bucket-region", "", "AWS region of the S3 bucket to store etcd backups.")
+
+	cmd.MarkFlagRequired("cluster-id")
+	cmd.MarkFlagRequired("etcd-endpoint")
+	cmd.MarkFlagRequired("s3-bucket-name")
+
+	if err := cmd.Execute(); err != nil {
+		os.Exit(1)
+	}
+}
+
+// go run etcd-backup/main.go etcd-backup --cluster-id mraee-hosted-cluster --etcd-endpoint localhost:2379 --s3-bucket-name mraee-hypershift-bucket
+func run(opts options) error {
+	tlsInfo := transport.TLSInfo{
+		CertFile:      "/etc/etcd/tls/client/etcd-client.crt",
+		KeyFile:       "/etc/etcd/tls/client/etcd-client.key",
+		TrustedCAFile: "/etc/etcd/tls/etcd-ca/ca.crt",
+	}
+
+	tlsConfig, err := tlsInfo.ClientConfig()
+	if err != nil {
+		return fmt.Errorf("error creating client TLSConfig: %w", err)
+	}
+
+	cfg := clientv3.Config{
+		Endpoints:   []string{opts.etcdEndpoint},
+		DialTimeout: DefaultDialTimeout,
+		TLS:         tlsConfig,
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), DefaultEtcdClientTimeout)
+	defer cancel()
+
+	filePath := "/tmp/snapshot.db"
+	err = snapshot.Save(ctx, zap.L(), cfg, filePath)
```

Comment:
I wonder if it's not better to take the snapshot using an etcd container instead of vendoring all of the code that is needed in the hypershift repo. The version of the code would be easier to maintain (by just updating the container image we use instead of remembering to update vendored code)

Two possibilities:
1) Change this utility to simply be a backup uploader and construct a job where an etcd container performs a backup to an emptydir in an init container and then the main container runs this code to upload the snapshot from the same dir

or 

2) Exec out to etcdctl snapshot save from here (would require that you make this binary accessible to the etcd container so that it can be invoked in a container where both etcdctl and this binary are present)

Reply:
I would discard option 2 because of the additional network journey and the image building overhead.

+1 for option 1, then if we ever need more control we can always get back to vendoring.

Reply:
We are going to vendor the etcd-client code also for the etcd-defrag controller https://github.com/openshift/hypershift/pull/2869 so this wouldn't actually add any new dependecnies.

Reply:
I don't think we want to do that for the defrag controller, but rather do the same than in 1 here? thoughts cc @csrwng

Reply:
implemented option 2 as discussed.